### ISSUE TITLE
Create Default Logging level configuration for ESAPI library

### ISSUE THREAD

**Issue by jeremiahjstacey** [ESAPI/esapi-java-legacy#631]
Derived from issue #630, the default logging level of all ESAPI loggers is `Level.ALL`.
This should be moved to a configurable component in ESAPI.proprites and all of the LoggerFactories should be updated to use that value for generating new Logger instances.

**Comment by kwwall** [comment]
@jeremiahjstacey - It probably was an oversight not to allow the default logging level not to be specified in the **ESAPI.properties** file, but I do think we need to be careful about the implementation. For example, if we add something like `Logger.DefaultLogLevel` then:

- How does that interact with a default log level found in a configuration file specific to the type of logger (e.g., SLF4J, JUL) if they are set differently there? At a minimum, we would have to describe in detail how they might conflict.
- How does that affect calls to `Logger.always()`, which should _always_ log (unless of course the logging were disabled, but setting the default log level to 'OFF'.
- Should we even _allow_ 'OFF' (which would disable ESAPI logging) as an option to be set from the ESAPI.properties file?

I'm not sure that all of these have obvious answers to them which is why I am raising them now for discussion.

**Comment by jeremiahjstacey** [comment]
@kwwall, I fully agree.  That is why I pushed this to another task.

I believe we did discuss this a bit in the initial implementation and had the same concerns with a condensed timeline for the release.  I think the improvement got lost in the shuffle.

We should work through the cases you've specified before starting work on this to ensure a solid understanding of the desired outcome.

**Comment by jeremiahjstacey** [comment]
Something else to note, that may or may not matter -- Right now the ESAPI library does no independent filtering. We did this to avoid the possible points of confusion resulting from the combined set of logging between ESAPI and the underlying logging implementation in the client environment.

At this point, ESAPI is not filtering log messages and expects that the client's preferred logger is configured in a way that will provide the desired output for the environment.

This is the precise reason why all of the LogFactory implementations default to ALL as the configured logging level.

Out of simplicity, I almost want to suggest that we not support a configuration setting for altering the default filtered level of the ESAPI library.  The maintenance cost of exposing this debug path as an expected use case sounds costly to me.

If someone wants to override the existing configuration of ESAPI's "Forward Everything" behavior, then this can be achieved by creating a new implementation of the LogFactory in the client's environment and updating the LoggerFactory property in ESAPI.properties.

They own the custom factory, they own the debugging.  We allow extension, but I don't think I want to accept responsibility for this use case as there are too many "How is your environment configured?" factors that would need to be accounted for during diagnostics.

To summarize:

1) The current configuration enables all filtering to be done in the client's preferred logging implementation
2) If log filtering is required at the ESAPI level then a client can create a new custom factory and set it in the ESAPI.properties of their own environment.

I'm not sure this work is worth doing once we have corrected the logic check in #630.

**Comment by sjoshid** [comment]
@jeremiahjstacey created a PR. Just for discussion. Im sure there might be more consequences and indepth discussion might be required.

**Comment by jeremiahjstacey** [comment]
7/19/2021  Email was sent to Esapi users and developers for feedback on interest of this feature.
As of 8/2/2021, no feedback received to either my email or on this task.

@kwwall  @xeno6696  Have either of you heard anything since the request for input?

**Comment by xeno6696** [comment]
I've gotten nothing.

**Comment by kwwall** [comment]
Nada! Although,  TBH, I'm not surprised. Probably should rattle the cage
one more time for good measure.

-kevin

**Comment by jeremiahjstacey** [comment]
**Email sent to ESAPI Users and ESAPI Dev mailing list on 7/19/2021**

> -SUMMARY-
> 
> Presently, the ESAPI Loggers default to forwarding every single log request to the delegate logging implementation, and rely on the client's logger-specific configuration to appropriately handle that event as desired for that environment.  In a recently filed issue it was brought up that there is not currently a method by which a client can modify the default ESAPI logging level.  It is hard-coded to forward all messages.
> 
> Effectively, what we're talking about is enabling a new level of log filtering that would be applied in the ESAPI Logger wrapper before the delegate logging implementation is notified of a logging event.
> 
> For further details, see https://github.com/ESAPI/esapi-java-legacy/issues/631
> 
> We are specifically looking for input/feedback into the desire for this capability in the currently supported ESAPI Logging Implementations: SLF4J, Java, and Log4J (deprecated).
> 
> -PROS-
> 
> If we move forward with exposing this configuration to pre-filter the log events, then the clients will have a bit more control of what logs are propagated from ESAPI into the delegate logging implementation, and that may result in some performance enhancements for avoiding the lookup on the additional information that ESAPI prepends to log events.
> 
> -CONS-
> 
> This capability comes at the cost of potentially having to configure and maintain the two logging implementations, and also understand the impacts of the logging levels and associations between the ESAPI library and the client's configured logger.  There are also some technical hurdles that will need to be resolved to reduce the possibility of the creation of incompatible configurations being created.
> 
> Some of those hurdles have been identified by kwwall already:
> 
> https://github.com/ESAPI/esapi-java-legacy/issues/631#issuecomment-881929323
> 
>     How does that interact with a default log level found in a configuration file specific to the type of logger (e.g., SLF4J, JUL) if they are set differently there? At a minimum, we would have to describe in detail how they might conflict.
>     How does that affect calls to Logger.always(), which should always log (unless of course the logging were disabled, by setting the default log level to 'OFF').
>     Should we even allow 'OFF' (which would disable ESAPI logging) as an option to be set from the ESAPI.properties file?
>     If we proceed with this, should we change the meaning of Logger.getESAPILevel() to return the (say) Logger.DefaultLogLevel in ESAPI.properties or should it return the log level of what was set with Logger.setLevel() (which would apply to the specific Logger instance) or the default of what is in the corresponding underlying logger's configuration file(s) (assuming that is even feasible without jumping through too many hoops)?
> 
> Other concerns are: Is there something else that we are overlooking? What are the desired use cases that ESAPI needs to support?
> 
> -FEEDBACK-
> 
> This email chain is the preferred method to engage in discussion.  If any are uncomfortable with responding to the group, then please feel free to reach out to me individually and we can work together to capture your input.  I'm sure kwwall or xeno6696 would also be available for discussion if any prefer those channels.
> 
> 
> I appreciate your time, and any feedback regarding this potential extension to the ESAPI library.
> 
> -Jeremiah Stacey
> 

**Follow-up sent on 8/13/2021**

> Been about a month and there has been no input on this topic.
> 
> Unless someone reaches out in favor of this capability by the end of the month, I plan to recommend to the project leads to simply reject this as a 'wont fix' item at this time.  If a need or use case comes up in the future we can certainly re-assess; however, I would much prefer not to incur this maintenance cost on the project unless it would fulfill a specific and immediate need.
> 
> -Jeremiah

**Comment by gpfountz** [comment]
I vote this be fixed.  log.isDebugEnabled() is always returning true even when the class is configured for info level logging - I cannot see any way that it is ok for this method to always return true.  Example: we have code inside our log.isDebugEnabled() code block that *MUST* only run when I have that class set for debug level logging.

**Comment by jeremiahjstacey** [comment]
@gpfountz that problem has been addressed in #630, and has already been addressed in the current baseline.

**Comment by gpfountz** [comment]
Thank you!

**Comment by jeremiahjstacey** [comment]
Moving forward with previously mentioned 'wontfix' resolution of this effort
